pkg/kv/kvserver/replica_init.go

```diff
@@ -71,13 +71,29 @@ func newInitializedReplica(store *Store, loaded kvstorage.LoadedReplicaState) (*
 
 // newUninitializedReplica constructs an uninitialized Replica with the given
 // range/replica ID. The returned replica remains uninitialized until
-// Replica.loadRaftMuLockedReplicaMuLocked() is called.
+// Replica.initRaftMuLockedReplicaMuLocked() is called, but has a temporary Raft
+// group that can be used e.g. for elections or snapshot application.
 //
 // TODO(#94912): we actually have another initialization path which should be
 // refactored: Replica.initFromSnapshotLockedRaftMuLocked().
 func newUninitializedReplica(
 	store *Store, rangeID roachpb.RangeID, replicaID roachpb.ReplicaID,
-) *Replica {
+) (*Replica, error) {
+	r := newReplica(store, rangeID, replicaID)
+	r.raftMu.Lock()
+	defer r.raftMu.Unlock()
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if err := r.initRaftGroupRaftMuLockedReplicaMuLocked(); err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+// newReplica creates a new, bare replica without a Raft group. Use either
+// newInitializedReplica() or newUninitializedReplica() instead, which will
+// include an appropriate Raft group.
+func newReplica(store *Store, rangeID roachpb.RangeID, replicaID roachpb.ReplicaID) *Replica {
```

Comment:
This seems somewhat worse than before, we already need uninited replicas to have the ability to become initialized, and both are valid concepts whereas what's returned from `newReplica` is essentially a "broken" replica. Were there complications in the old structure that made this `newReplica` necessary?

Reply:
This is an internal helper for `newInitializedReplica()` and `newUninitializedReplica()`. If we have `Initialized` simply call through to `Uninitialized` then we'll unnecessarily create a Raft group twice in rapid succession, because `Initialized` has to replace the Raft group from an uninitialized replica.

We could alternatively pass a boolean parameter to `Uninitialized` controlling whether to create a Raft group, but that seemed even worse since `Uninitialized` is a widely used API, which could lead callers to believe that it was legit to create it with a Raft group.

Reply:
I renamed it to `newUninitializedReplicaWithoutRaftGroup` to try to make this a bit clearer, but didn't find a good way to otherwise avoid this.